### DISCUSSION TITLE
Preset Saving String as Text File

### DISCUSSION THREAD

**Discussion by puffer3** [RomanKubiak/ctrlr#448] [Q&A] [answered]
Hi All,

I am writing an OB8 panel and trying to implement some patch saving functions.  

My goal is to have: 

a listbox containing saved patches
a button to set the root folder for where the patches are stored
a dialog box for the user to enter the patch name or displays current patch name 
and an Initialized patch load button

The patch section is under construction, but I have lots of the rest of the interface working well. 

![Screen Shot 2022-03-15 at 2 44 02 PM](https://user-images.githubusercontent.com/6006231/158478786-1e0f117c-3b62-42de-b051-67a97dc2b2c1.png)

So far I have been able to make a table of modulators and print the values to the console.  I am trying to use the saveDataToText examples from the file panel but replaceWithText is giving me a C error.  it might be because I am on an older laptop or dont have some Xcode properly installed or something.  Any tips?  Here's my method thats an amalgamation of the Files Example panel and some reference to the DX-11 panel. 

![Screen Shot 2022-03-15 at 2 17 51 PM](https://user-images.githubusercontent.com/6006231/158477808-a978a71f-5d0a-4b42-ab58-e5dc4874f1bf.png)

```

function ListTest()

PatchDataTable = {}

patch1 = panel:getModulatorByName("MproVol"):getModulatorValue()
PatchDataTable[1] = (patch1)
patch2 = panel:getModulatorByName("voiceDetune"):getModulatorValue()
PatchDataTable[2] = (patch2)
patch3 = panel:getModulatorByName("CportT"):getModulatorValue()
PatchDataTable[3] = (patch3)
patch4 = panel:getModulatorByName("Unison"):getModulatorValue()
PatchDataTable[4] = (patch4)
patch5 = panel:getModulatorByName("O2Detune"):getModulatorValue()
PatchDataTable[5] = (patch5)
patch6 = panel:getModulatorByName("LFOfrq"):getModulatorValue()
PatchDataTable[6] = (patch6)
patch7 = panel:getModulatorByName("LFOwave"):getModulatorValue()
PatchDataTable[7] = (patch7)
patch8 = panel:getModulatorByName("Mdepth1"):getModulatorValue()
PatchDataTable[8] = (patch8)
patch9 = panel:getModulatorByName("ModFreqO1"):getModulatorValue()
PatchDataTable[9] = (patch9)
patch10 = panel:getModulatorByName("ModFtrFreq"):getModulatorValue()
PatchDataTable[10] = (patch10)
patch11 = panel:getModulatorByName("ModFreqO2"):getModulatorValue()
PatchDataTable[11] = (patch11)
patch12 = panel:getModulatorByName("Mdepth2"):getModulatorValue()
PatchDataTable[12] = (patch12)
patch13 = panel:getModulatorByName("ModPwmO1"):getModulatorValue()
PatchDataTable[13] = (patch13)
patch14 = panel:getModulatorByName("ModPwmO2"):getModulatorValue()
PatchDataTable[14] = (patch14)
patch15 = panel:getModulatorByName("ModVol"):getModulatorValue()
PatchDataTable[15] = (patch15)
patch16 = panel:getModulatorByName("O1freq"):getModulatorValue()
PatchDataTable[16] = (patch16)
patch17 = panel:getModulatorByName("O1Wave"):getModulatorValue()
PatchDataTable[17] = (patch17)
patch18 = panel:getModulatorByName("OscPW"):getModulatorValue()
PatchDataTable[18] = (patch18)
patch19 = panel:getModulatorByName("O1sync"):getModulatorValue()
PatchDataTable[19] = (patch19)
patch20 = panel:getModulatorByName("O1ftrenv"):getModulatorValue()
PatchDataTable[20] = (patch20)
patch21 = panel:getModulatorByName("O2freq"):getModulatorValue()
PatchDataTable[21] = (patch21)
patch22 = panel:getModulatorByName("O2wave"):getModulatorValue()
PatchDataTable[22] = (patch22)
patch23 = panel:getModulatorByName("ftrFreq"):getModulatorValue()
PatchDataTable[23] = (patch23)
patch24 = panel:getModulatorByName("ftrRez"):getModulatorValue()
PatchDataTable[24] = (patch24)
patch25 = panel:getModulatorByName("ftrMod"):getModulatorValue()
PatchDataTable[25] = (patch25)
patch26 = panel:getModulatorByName("O1Fon"):getModulatorValue()
PatchDataTable[26] = (patch26)
patch27 = panel:getModulatorByName("O2on"):getModulatorValue()
PatchDataTable[27] = (patch27)
patch28 = panel:getModulatorByName("FtrNoise"):getModulatorValue()
PatchDataTable[28] = (patch28)
patch29 = panel:getModulatorByName("FtrKeytrack"):getModulatorValue()
PatchDataTable[29] = (patch29)
patch30 = panel:getModulatorByName("Ftr4pole"):getModulatorValue()
PatchDataTable[30] = (patch30)
patch31 = panel:getModulatorByName("EnvAmpA"):getModulatorValue()
PatchDataTable[31] = (patch31)
patch32 = panel:getModulatorByName("EnvAmpD"):getModulatorValue()
PatchDataTable[32] = (patch32)
patch33 = panel:getModulatorByName("EnvAmpS"):getModulatorValue()
PatchDataTable[33] = (patch33)
patch34 = panel:getModulatorByName("EnvAmpR"):getModulatorValue()
PatchDataTable[34] = (patch34)
patch35 = panel:getModulatorByName("EnvFtrA"):getModulatorValue()
PatchDataTable[35] = (patch35)
patch36 = panel:getModulatorByName("EnvFtrD"):getModulatorValue()
PatchDataTable[36] = (patch36)
patch37 = panel:getModulatorByName("EnvFtrS"):getModulatorValue()
PatchDataTable[37] = (patch37)
patch38 = panel:getModulatorByName("EnvFtrR"):getModulatorValue()
PatchDataTable[38] = (patch38)
patch39 = panel:getModulatorByName("P2atkm1"):getModulatorValue()
PatchDataTable[39] = (patch39)
patch40 = panel:getModulatorByName("P2Dcym1"):getModulatorValue()
PatchDataTable[40] = (patch40)
patch41 = panel:getModulatorByName("P2atkm2"):getModulatorValue()
PatchDataTable[41] = (patch41)
patch42 = panel:getModulatorByName("P2Dcym2"):getModulatorValue()
PatchDataTable[42] = (patch42)
patch43 = panel:getModulatorByName("P2envMod"):getModulatorValue()
PatchDataTable[43] = (patch43)
patch44 = panel:getModulatorByName("P2retrig"):getModulatorValue()
PatchDataTable[44] = (patch44)
patch45 = panel:getModulatorByName("P2LFOtrack"):getModulatorValue()
PatchDataTable[45] = (patch45)
patch46 = panel:getModulatorByName("P2TrigP"):getModulatorValue()
PatchDataTable[46] = (patch46)
patch47 = panel:getModulatorByName("P2Qpwm"):getModulatorValue()
PatchDataTable[47] = (patch47)
patch48 = panel:getModulatorByName("P2Qfm"):getModulatorValue()
PatchDataTable[48] = (patch48)
patch49 = panel:getModulatorByName("P2inv1"):getModulatorValue()
PatchDataTable[49] = (patch49)
patch50 = panel:getModulatorByName("P2inv2"):getModulatorValue()
PatchDataTable[50] = (patch50)
patch51 = panel:getModulatorByName("LFOfm"):getModulatorValue()
PatchDataTable[51] = (patch51)




--print test of values to console
x = 1
for i = 1,51 do
console (string.format(PatchDataTable[x]))
x = x + 1
end


--get name of patch from label window
Name2File = panel:getModulatorByName("Patch_Name"):getComponent():getProperty("uiLabelText")

--create string concatenated from tabel
file_complete = table.concat(PatchDataTable, " ", 1, 51)


--print Patch Data
console ("Patch to Save is: "..file_complete)



-- Create a dialog window for saving file

-- Get current root location
currentRootFolder = panel:getModulatorByName("File_Browser"):getComponent():getProperty("uiFileListCurrentRoot")

-- Suggest current patch name without space
suggestPatchName = Name2File:gsub("%s+", "")

-- Suggest patch name and current folder
suggestPatchNameFolder = currentRootFolder.."/"..suggestPatchName
	
fileToWrite = utils.saveFileWindow("Save patch to disk", File(suggestPatchNameFolder), "*.txt", true)

if fileToWrite:isValid() == false then
return
end

-- Check if the file exists
if fileToWrite:existsAsFile() == false then
		
-- If file does not exist, then create it
if fileToWrite:create() == false then

-- If file cannot be created, then fail here
utils.warnWindow ("\n\nSorry, the Editor failed to\nsave the patch to disk!", "The file does not exist.")

return
end
end

	if fileToWrite:replaceWithText (file_complete, false, false) == false then
		utils.warnWindow ("File write", "Failed to write data to file: "..fileToWrite.getFullPathName())
	end


end

```

I am sure there is an easier way to collect the modulators but I am simply trying the clunkiest most reliable way possible.  I will still have to parse out the data and apply it to modulators when I figure out a load method, but for now I need to fill this text file.

some of my control names are silly.  This was my first panel ever. I am doing a little better after the jupiter8.  We will see how I do on the Matrix12.

**Comment by Godlike-Productions** [comment]
Change this line 

`if fileToWrite:replaceWithText (file_complete, false, false) == false then`

with

`if fileToWrite:replaceWithText (file_complete, false, false, "\n") == false then`

Depending on which version of ctrlr you are using the \n may or may not be required.

More discussion here

https://github.com/RomanKubiak/ctrlr/discussions/196#discussioncomment-420291

**Reply by puffer3** [reply]
This worked! 

As you have probably gathered I am not a very experienced programmer.  Does this format of numbers have a severe disadvantage to .json?  I am trying to get something working quickly and plan on implementing .json after I can learn it more.

**Reply by Godlike-Productions** [reply]
Which format or numbers?  The ones in your table?  Numbers should be converted automatically by the json4lua encode and decode functions.  Those are under the Built-In section of the LUA Editor.  MemoryBlock's (binary data) need to be converted, but they are more efficient, and easier to send out as sysex, and have a few other advantages when it comes to bit and byte manipulation.

If you store modulator values in a table, you will need to convert them to hex strings, and concatenate them to be able to form a sysex message.

**Comment by puffer3** [comment]
I see why saving the numbers as a long string with spaces in between is not the best idea.  I was hoping to forgo sysex and just reassign to the modulators and send a snapshot.  I am having trouble with file loading due to lack of documentation and my own ignorance.

Most of what I have seen of .json examples are for full banks.  Are there good examples of saving single patches I am not turning up?  

People must have worked out how to save single patches as files somewhere, but I have not found it.

**Reply by Godlike-Productions** [reply]
Unfortunately there isn't a simple solution to this. It depends on how you format stuff for saving and making sure you have the structure to load it into.

A lot of it has to do with how you pull the data into ctrlr.  Normally you do a dump request.  The data comes in as binary and gets stored as userData (essentially binary data).  You can then save that into a memory block, or you convert it into a hex string.  You can then do manipulation on the data and send it out or save it.

For the OB8 panel, creating a patch from the panel, assuming it uses standard 127 value messages you need to do something like this.

Note, you will need to make sure that you have defined the sysex header and termination within the midiBuffer somewhere.  I build the string dynamically, but for now, you could just stick fixed values into the table..  ie midiBuffer[1] = 0xf0...

Anyway, here's some simplified code that should build a memoryblock that you can then prepare for saving to json.

```
env1DecaySliderValue = panel:getModulatorByName("env1DecaySlider"):getValue()
midiBuffer[20] = env1DecaySliderValue

-- do this for each Modulator and stick it into the correct position of midiBuffer, which is a table of integers

--Create midiMemoryBlock,
	
        midiBufferMb:createFromTable(midiBuffer)
	
        

```

You could leave it as a table and export to .json, but there are bugs with converting a table to a MIDI Message in some versions of ctrlr, so keeping it as memoryblock is good practice.

OK, next, I do an encoding step.  I do it for a bank, but the important bit is this line

```
sysexString = bankToEncode[i].sysex:toHexString(1)
```

I then build a patch (I do this in a for loop for each patch in a bank, but this step builds a bank of 1 patch.  For details on how I build the patch and bank classes see this post - https://github.com/damiensellier/CtrlrX/wiki/Classes-for-Patch-Storage

```
i=1
tempBank[i] = patch:new({name=nameString, author=authorString, category=categoryString, sysex=sysexString})
```

Then I encode the json

```
local encodedBank = json.encode(tempBank)
    return encodedBank
```

Then save the file as you normally do, but save it as text.

For decoding, you do it the opposite way.  Load the .json file using the open file dialog.  This will build the class table for you.  All you need to do is convert the sysex string to a memoryblock, which is simple.

Here's my entire function 
```
function decodeBank(jsonToDecode)
		-- This function takes a json String previously encoded with encodeBank, and returns a patchBank

    
    if not isPanelReady() then -- this is a checking function that makes sure that ctrlr is ready to run.  You'll need to make this function yourself.  Leave the if out for testing.
		return
	end --if
     
   
    local decodedBank = json.decode(jsonToDecode)

    for i,v in ipairs (decodedBank) do
        v.sysex = MemoryBlock(v.sysex)
    end --for  
    return decodedBank
end
```


If you just want to save the sysex string, you can just save the variable to a binary file, and then reopen it as binary, which is how you export and import sysex files.

**Comment by dobo365** [comment]
Hi,
  I'm doing this in all my panels... Provided your info is in a string variable sTemp, simply do:

```
	-- Save the displayed info to a file
	f = utils.saveFileWindow ("Save program information to file", File(""), "*.txt", true)
	if f:isValid() then
		if f:existsAsFile() == false then
			f:create()
		end

		-- Write introduction text based on the Info type
		sTemp = "\n*** PROGRAM INFORMATION\n*** DATE: "..os.date("%Y-%m-%d").."\n\n"
--		sTemp = sTemp.."----------------------------+-----------------------------+-----------------------------+\n"
		f:replaceWithText(sTemp, false, false)

		-- Feedback to user
		utils.infoWindow("Save program information to file", "Program information saved in \n"..f:getFullPathName())
	end
```

As I said many times as well, it is better to initialize all your mods (panel:getModulatorByName() in a separate method AssignModulators() that you call with PanelLoaded(). Then you refer to the mod variables instead of calling panel:getModulatorByName() each time.
For bank mgmt, simple storage in MemoryBlock then just use getByte() or getRange() to extract the info you need.
External storage by writing the memoryblock in a file as sysex.
Feel free to look at my Pro-2 panel: [https://www.sunnysynths.com/dsi-pro-2](https://www.sunnysynths.com/dsi-pro-2)

**Comment by puffer3** [comment]
ok so my panels are saving to files just fine.  If I want to add a name via a uiLabel is it as simple as adding this code to the init function

init = function(--[[ CtrlrInstance --]] type)
    FEXT = "json"
patchName=panel:getModulatorByName("patchName")


and a variable to the patch section:

patchDataTable = {
"patchName",


or is it more complicated because it is a static label with a string and not a modulator with a numeric value?  Its on the modulator list, but so far this code isnt adding my name back in.  



[OB8patchNameTest.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8525180/OB8patchNameTest.bpanelz.zip)

**Comment by dnaldoog** [comment, marked as answer]
This version includes the previous hover blocker and also I have changed the presetNames table to a hash array. Each key has a value "value","label" - if it is a "value" then the json file grabs the modulator value. If it is a label "label" the json file gets the label text. You could extend this to grab say "getValueMapped" by having another value lookup "mapped" etc etc but don't worry about that for now.

<pre>
patchDataTable = { 
voiceDetune="value",
progVolume="value",
lfoModDpth1="value",
...
osc1FreqInvert="value",
customPatchName="label",
}-- //end patchDataTable
</pre>

so now the json code looks like this:

<pre>
 local t = {}
    for k, v in pairs(patchDataTable) do
	if v=="value" then
        t[k] = _G[k]:getModulatorValue()
	elseif v=="label" then
 		t[k] = L(panel:getLabelComponent(k):getText())
	end
    end
    json_str = json.encode(t)
</pre>


I also included a custom _function blocker class_, which should work in standalone and vst.

The standard `panel:getBootstrapState()` sometimes blocks lua in my experience.

[OB8_4_0_OB-8DAGDD.zip](https://github.com/RomanKubiak/ctrlr/files/8526250/OB8_4_0_OB-8DAGDD.zip)

**Reply by puffer3** [reply]
Thanks for your help!

**Reply by puffer3** [reply]
@dnaldoog 

I am trying implement the patch saving functions we developed here on another panel (prophet 10) to understand them better.  I cant seem to get it working after some hours of trying.  I have been running the functions in the console and it seems if I comment out the line 

        t[k] = _G[k]:getModulatorValue()

in the save_as_file function I can get a window to appear.  It seems that my patchDataTable isnt getting created or the panel is getting blocked? I am not sure which.  humbly asking for help.  I included the latest OB8 panel version that I took the code functions from.

**Reply by dnaldoog** [reply]
Hi @puffer3 ,

Can you post the panel in question for the prophet 10?

Regards,

John.

**Reply by puffer3** [reply]
yes sorry it seems my uploads are acting funny today!
[Prophet10PathSaveExport.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8596184/Prophet10PathSaveExport.bpanelz.zip)

**Reply by dnaldoog** [reply]
Hi @puffer3 

The problem was you had 4mod names (with the same name) `lfoOsc2FM="value",` in the `patchDataTable` but there are no modulators with those names. There are however `lfoOsc2FM-21, 26,27 & lfoOsc2FM-30`, so they need to be added to the table, but you would need to substitute the '-' for '_' or some other naming scheme to make the name legal as a _lua_ variable name. Note that you were declaring each modulator in the init function and then relisting again in `patchDataTable` - this is doubling up, so it would be better to use `patchDataTable` to initialise: 

<pre>
for k in pairs (patchDataTable) do
_G[k]=panel:getModulatorByName(k)
end
</pre>
[Prophet 10 HJG_0.0_10_JG altered 2022-04-30.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8596589/Prophet.10.HJG_0.0_10_JG.altered.2022-04-30.panel.zip)

----
### But forget all that!
There's a simpler way. Because caching modulator names is not necessary when saving a file (it might be necessary when constantly sending out large sysex strings to a synth etc), there is no need for all that `_G[k]` stuff. Instead in this panel we just look up each value using `panel:getModulatorByName(k):getModulatorValue()` when saving the file. To make a variable name with a dash (or anything in it) legal as a lua table key, we surround it by `[" "]` so we can just add `["lfoOsc2FM-21"]="value",`  in  "patchDataTable". Note that in `init`, there is no initialisation of modulators now. It's probably better and simpler! **I recommend you use this panel, not the one above!**

[Prophet 10 HJG_0.0_10_NON LUA VARIABLE_2022-04-30_18-30.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8596594/Prophet.10.HJG_0.0_10_NON.LUA.VARIABLE_2022-04-30_18-30.panel.zip)

----
<pre>
local t = {}
    for k, v in pairs(patchDataTable) do
	if v=="value" then
       t[k] = panel:getModulatorByName(k):getModulatorValue()
	elseif v=="label" then
 		t[k] = L(panel:getLabelComponent(k):getText())
	end
    end
    json_str = json.encode(t)
</pre>

**Reply by puffer3** [reply]
ugh its still not working for me! I am having a hard time troubleshooting i am terrible at programming.  Those illegal modulator names were gumming up the works because I was copying and pasting when building the UI. The save window isnt showing up and I am stuck!

**Reply by dnaldoog** [reply]
> ugh its still not working for me! I am having a hard time troubleshooting i am terrible at programming. Those illegal modulator names were gumming up the works because I was copying and pasting when building the UI. The save window isnt showing up and I am stuck!

Are you using the second program I posted? You can't cut and paste from it to another version you have been working on; you need the whole program. Or cut and paste from your version into this one.

I don't think you are `terrible at programming`. Programming is very difficult (for me at least). Keep persevering - it takes time.

[Prophet 10 HJG_0.0_10_NON LUA VARIABLE_2022-04-30_18-30.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8608306/Prophet.10.HJG_0.0_10_NON.LUA.VARIABLE_2022-04-30_18-30.panel.zip)

**Reply by puffer3** [reply]
It seems I am still not getting the save window to open.  How can I troubleshoot it? 

I am using your newest panel.

**Comment by dnaldoog** [comment]
You've got two 'save' buttons.

One of them has the button value `Value one=127` set. For some reason this string seems to generate a `source=6` parameter being passed to the function `save_to_file`.  That means the function won't fire. It's probably a bug. If you delete that _string_ the button works. The other save button was working.

The function requires  `source==4` to stop windows opening when you load two instances or load another panel in the same ctrlr session.

<pre>
save_as_file = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
	if myPanel:getBootstrapState() then
			return
	end
	if source ~=4 then return end -- you could delete this line
    local t = {}
...
</pre>

![savebutton](https://user-images.githubusercontent.com/76831683/166588546-630e85db-2bcd-4a12-9356-28b2f0b2dc18.png)

**Reply by puffer3** [reply]
Ah Yes thank you! That must have been the problem! I am learning all these little issues thanks to you.

**Reply by puffer3** [reply]
Okay I cleaned up the two panels we are close to finishing (OB8 and Prophet 10) 

One question I have is:  Is it possible to send a snapshot from the load_as_file Lua script?  I have several threads open from the old forum about snapshots but none of them have the syntax for that.

Would it be possible to set the folder for the patch storage and use navigation buttons on either side of the patch name box to cycle thru them?  Is this something someone has done before? 

I had these saving great on mac but when I loaded them onto the target use PC the save dialogs wouldnt open.  I went thru and cleared out any garbage modulators from my early inexperienced UI layout.  Is there anything else that could be impeding the save from working in Frutiy Loops as a VST on a windows PC?

Here are my two panels in their current state:
[Prophet 10 HJG _0.0_14_NON LUA VARIABLE_2022-05-06_15-59.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8644019/Prophet.10.HJG._0.0_14_NON.LUA.VARIABLE_2022-05-06_15-59.bpanelz.zip)
[OB8_4_2_OB-8DAGDD_2022-05-06_15-58.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8644046/OB8_4_2_OB-8DAGDD_2022-05-06_15-58.bpanelz.zip)

**Reply by dnaldoog** [reply]
> One question I have is: Is it possible to send a snapshot from the load_as_file Lua script? I have several threads open from the old forum about snapshots but none of them have the syntax for that.

Hi @puffer3 ,

To send a snapshot of the loaded _json_ file to the synth just comment out the `panelMidiPauseOut` code:
<pre>
load_file = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
	if myPanel:getBootstrapState() then
        return
    end
   -- panel:setPropertyInt("panelMidiPauseOut", 1) -- !PAUSE SENDING MIDI
    fileToRead =
        utils.openFileWindow("Open json file", File.getSpecialLocation(File.userDesktopDirectory), "*." .. FEXT, true)
    local obj = ""

    if fileToRead:existsAsFile() then
        local str = fileToRead:loadFileAsString()
        obj = json.decode(str)
    end
    if type(obj) ~= "table" then
        return
    end -- bail out if something's not right

    for k, v in pairs(obj) do
        if patchDataTable[k] ~= nil then
            if patchDataTable[k] == "value" then
                panel:getModulatorByName(k):setModulatorValue(v, false, true, false)
            elseif patchDataTable[k] == "label" then
                panel:getLabelComponent(k):setText(v)
            end
        end
    end -- 	if patchDataTable[k] ~= nil then
  --  panel:setPropertyInt("panelMidiPauseOut", 0) -- !ALLOW SENDING MIDI
end
</pre>

**Reply by puffer3** [reply]
@dnaldoog Thanks John I should have seen the comments you had placed in there!

I dont actually have a prophet 10/ ob8 my friend does so I am using the assets and work we have done to make a controller for my modest prophet 600 gligli and OBSX in order to get better at fine tuning these panels.  I have spent some time copying the file saving code and cleaning up the modulators for the p600 but there is still something stopping me from saving and loading.  Can you take a look?  I have been hacking at it for an hour.
[Prophet 600 gligli_0.0_11_saving_2022-05-07_09-48.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8645606/Prophet.600.gligli_0.0_11_saving_2022-05-07_09-48.bpanelz.zip)

**Reply by dnaldoog** [reply]
Hi @puffer3 

That panel was missing the hidden button that unblocks the panel on load.

It works like this:

The `myPanel:getBootstrapState()` method is set to **TRUE** when the panel loads, so no functions will work. This stops functions from running on start - dialog boxes popping up etc. 

There is a button that is fired on start up, but it's hidden. It contains a timer that sets `myPanel:getBootstrapState()`  back to **FALSE** which then allows functions to run. That delay in the timer stops functions from running, because the panel has now loaded by the time the timer has set it to **FALSE**. The reason the button (that is hidden) fires on startup is because it doesn't contain the code `if myPanel:getBootstrapState()== true then return end` in it so it fires off the timer when the panel loads.

This `myPanel:getBootstrapState()` object is a substitute version for Ctrlr's `panel:getBootstrapState()` which, in my experience, sometimes doesn't work in standalone/VST. https://ctrlr.org/forums/topic/standalone-images-not-loading/

![unblockbutton](https://user-images.githubusercontent.com/76831683/167274338-5e7c7674-cdbd-4656-886c-9309e5157bde.png)

[Prophet 600 gligli (with unblock button)_0.0_11_saving_2022-05-08_07-41.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8646067/Prophet.600.gligli.with.unblock.button._0.0_11_saving_2022-05-08_07-41.panel.zip)

**Reply by puffer3** [reply]
of course I forgot to copy the invisible button! 

My savior @dnaldoog

**Reply by dnaldoog** [reply]
> of course I forgot to copy the invisible button!
> 
> My savior @dnaldoog

😂 ― It's tricky - the whole thing, but I believe that code represents a good template to build panels from. That's what I do now. Kind of like what you are doing. Get one panel down and copy that to create new ones!

**Comment by puffer3** [comment]
Alright so things still weren't working so I made the invisible button visible and then at least the dialogue came up. 

but now i am getting a C error ugh
![Screen Shot 2022-05-08 at 12 05 43 AM](https://user-images.githubusercontent.com/6006231/167285807-55072e31-44b3-4d27-83bb-81cead65a2e6.png)

**Reply by dnaldoog** [reply]
You've done something I can't duplicate. Can you post the panel that throws this error?

The unblock panel button should never need to be used and certainly not be visible. It is only fired when the panel starts and doesn't need to be visible at that. But whether it's visible or not shouldn't be causing the panel to throw that error. 🤔

**Reply by puffer3** [reply]
Hi @dnaldoog ,

This was using the same panel as you had posted.  When loading it seemed that the panel was still blocked so I made the button visible to see if clicking it would help.  I tied on my high sierra mac laptop and mac mini.  Are you using a PC?  

I tried with both VST and AU versions on my Macmini with 5.6.23 and macbook pro running 5.6.23  and got the same C error !

**Reply by dnaldoog** [reply]
I am using PC but I only ever use 5.3.201. For me it is stable and I know how to get around most of its bugs. And I have not heard of many issues with my panels from Mac users with my vst/au standalone builds using 5.3.198.

**Reply by puffer3** [reply]
Ok I will download that version and give it a try!

**Reply by dnaldoog** [reply]
> Ok I will download that version and give it a try!

I have them here in case the Ctrlr website goes down.

https://sourceforge.net/projects/ctrlr-utilities/files/Ctrlr/

**Reply by dnaldoog** [reply]
> > Ok I will download that version and give it a try!
> 
> I have them here in case the Ctrlr website goes down.
> 
> https://sourceforge.net/projects/ctrlr-utilities/files/Ctrlr/

I just tried 5.6.23 and it was crashing everywhere and not working for me on Windows! That is to say it is crashing on a couple of my panels developed with 5.3.201 ―maybe it works when developing a panel from scratch with that version? Dunno.

**Reply by puffer3** [reply]
I ran into this with some of @Godlike-Productions with some of the Lua syntax being a little different.  I can't remember exactly what but it was throwing a similar C error.  I tried back on 5.3.198 and it seems to be exporting.  I am going to try it in ableton next.  Thanks for the help!
